### PR TITLE
add graphs

### PR DESCRIPTION
**Pull request by mindas2000**
graph1

### CODE REVIEW

**Review by mindas2000 — COMMENTED** [review]
אין שינויים

File: client-side/src/graphs/graphs.jsx
Comment (mindas2000) [inline client-side/src/graphs/graphs.jsx:1]: הנתונים כרגע סטטיים יש צורך להעביר לנתונים מהשרת

**Review by maykelMarziano — APPROVED** [review]
GOOD

File: client-side/src/App.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import React, { useState } from 'react';
+import './App.scss';
+import DateTimePicker from './components/report.jsx'
+import VisitedWebsitesComponent from './graphs/graphs.jsx';
+function App() {
+
+  const [startDate, setStartDate] = useState('');
```
Comment (EA171) [inline client-side/src/App.js:7]: you shouldn't add code to the app, please move it to statistics component.
and also why do you have this App.js file, we have already App.jsx

File: client-side/src/App.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  const handleDateSubmit = (start, end) => {
+    setStartDate(start);
+    setEndDate(end);
+    setShowVisitedWebsites(true);
+
+  };
+
+  const user = {
+    name: "jack",
```
Comment (EA171) [inline client-side/src/App.js:19]: user shouldn't b hard-code it should come from redux

File: client-side/src/App.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    profileImage: "rimon.jpg",
+  }
+
+  return (
+    <div className="App">
+
+      <br></br>
+      <br></br>
+      <br></br>
+      <DateTimePicker onDateSubmit={handleDateSubmit} />
```
Comment (EA171) [inline client-side/src/App.js:32]: don't use br

File: client-side/src/graphs/graphs.jsx [hunk 4]
```diff
@@ -0,0 +1,313 @@
+import * as React from 'react';
+import { LineChart, ChartContainer, LinePlot, BarPlot, ChartsYAxis, ChartsXAxis, Gauge, PieChart, ChartsLegend } from '@mui/x-charts';
+import Stack from '@mui/material/Stack';
+import { useQuery, gql } from '@apollo/client';
+
+const GET_VISITED_WEBSITES = gql`
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:6]: please open a new file for all gql queries and aggregations.

File: client-side/src/graphs/graphs.jsx [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        },
+    },
+};
+
+
+const VisitedWebsitesComponent = ({ startDate, endDate, user }) => {
+    const websites = useQuery(GET_WEBSITE);
+    const visitedWebsite = useQuery(GET_VISITED_WEBSITES);
+    const users = useQuery(GET_USERS);
+    const colors = ["red", "yellow", "orange", "blue", "lightgreen"]
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:64]: colors should come from constants file

File: client-side/src/graphs/graphs.jsx [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    let dateStart = new Date(startDate.$d);
+    let year = dateStart.getFullYear();
+    let month = ('0' + (dateStart.getMonth() + 1)).slice(-2);
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:67]: 1. why do you add '0' in start, what will be when the month will be 10 for example? (same for the day)
2. for joining few strings and variables please use es6 normalization:
`0${dateStart.getMonth() + 1}`
3. why you need the slice(-2)?
4. you shouldn't write code inside the component, all the logic code should be inside functions

File: client-side/src/graphs/graphs.jsx [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+    let day = ('0' + dateStart.getDate()).slice(-2);
+    dateStart = `${year}-${month}-${day}`
+    let dateEnd = new Date(endDate.$d);
+    year = dateEnd.getFullYear();
+    month = ('0' + (dateEnd.getMonth() + 1)).slice(-2);
+    day = ('0' + dateEnd.getDate()).slice(-2);
+    dateEnd = `${year}-${month}-${day}`
+    if (websites.loading || users.loading) return <p>Loading...</p>;
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:75]: you shouldn't return here, all return states should be together

File: client-side/src/graphs/graphs.jsx [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+    if (websites.error  || users.error) return <p>Error</p>;
+
+    let webSite = websites.data.websites.map((obj, index) => ({ ...obj, color: colors[index] }))
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:79]: please use ? for case that the websites will be empty

File: client-side/src/graphs/graphs.jsx [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+    const u = users.data.users.find(u => u.email === user.email)
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:80]: please use ? for case that the users will be empty

File: client-side/src/graphs/graphs.jsx [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+    console.log({ u });
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:81]: remove console log from all your code

File: client-side/src/graphs/graphs.jsx [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+    const totalActivityTimeByWebsite = {};
+
+    u.visitsWebsites.forEach(visitedWebsite => {
+        const websiteName = visitedWebsite.websiteId.name;
+        const activityTime = visitedWebsite.visitsTime.reduce((total, visit) => {
+            let date = new Date(visit.visitDate)
+            year = date.getFullYear();
+            month = ('0' + (date.getMonth() + 1)).slice(-2);
+            day = ('0' + date.getDate()).slice(-2);
+            date = `${year}-${month}-${day}`
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:92]: please create a function for formatting date and use it in all places that you creating a date.

File: client-side/src/graphs/graphs.jsx [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+    dateEnd = `${year}-${month}-${day}`
+    if (websites.loading || users.loading) return <p>Loading...</p>;
+
+    if (websites.error  || users.error) return <p>Error</p>;
+
+    let webSite = websites.data.websites.map((obj, index) => ({ ...obj, color: colors[index] }))
+    const u = users.data.users.find(u => u.email === user.email)
+    console.log({ u });
+    
+    const totalActivityTimeByWebsite = {};
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:83]: shouldn't it be a state variable?

File: client-side/src/graphs/graphs.jsx [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+
+    webSite.forEach(site => {
+        const websiteName = site.name;
+        if (totalActivityTimeByWebsite[websiteName]) {
+            site.time = totalActivityTimeByWebsite[websiteName];
+        }
+    });
+    return (
+        <>
+            {/* <h2>שלפנו את נתוניך על פי החודש  {month}-{year}</h2> */}
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:114]: please remove comments

File: client-side/src/graphs/graphs.jsx [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+                    ]}
+                    {...otherProps}
+                />
+            </Stack>
+        </>
+
+    );
+
+}
+export default VisitedWebsitesComponent;
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:133]: please add prop-types to this component

File: client-side/src/index.js [hunk 15]
```diff
@@ -6,15 +6,28 @@ import reportWebVitals from './reportWebVitals';
 import { BrowserRouter } from 'react-router-dom';
 import { createTheme } from '@mui/material';
 import { ThemeProvider } from '@emotion/react';
+import { ApolloProvider, ApolloClient, InMemoryCache } from '@apollo/client';
 
 
+const client = new ApolloClient({
+
+uri: 'http://localhost:4000',
```
Comment (EA171) [inline client-side/src/index.js:14]: uri should be taken from env

**Review by EA171 — CHANGES_REQUESTED** [review]

File: client-side/src/components/statistics.jsx [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    const [showVisitedWebsites, setShowVisitedWebsites] = useState(false);
+
+    const handleDateSubmit = (start, end) => {
+        setStartDate(start);
+        setEndDate(end);
+        setShowVisitedWebsites(true);
+
+    };
+
+    const user = {
```
Comment (EA171) [inline client-side/src/components/statistics.jsx:18]: why this is hard-code?

File: client-side/src/graphs/graphs.jsx [hunk 17]
```diff
@@ -1,125 +1,92 @@
...
+        width: 400,
+        height: 200,
+        sx: {
+            [`.${ChartsLegend.root}`]: {
+                transform: 'translate(20px, 0)',
+            },
+        },
+    };
+   
+    VisitedWebsitesComponent.propTypes = {
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:75]: prop types should be defined outside the component, in the end of the file, before the return statment

File: client-side/src/graphs/graphs.jsx [hunk 18]
```diff
@@ -1,125 +1,92 @@
 import * as React from 'react';
+import PropTypes from 'prop-types'
 import { LineChart, ChartContainer, LinePlot, BarPlot, ChartsYAxis, ChartsXAxis, Gauge, PieChart, ChartsLegend } from '@mui/x-charts';
```
Comment (EA171) [inline client-side/src/graphs/graphs.jsx:3]: please remove all unnecessary imports

**Review by EA171 — CHANGES_REQUESTED** [review]

**Review by maykelMarziano — APPROVED** [review]

File: client-side/src/graphs/graphs.jsx [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+
+    const getRandomColor = () => `#${Math.floor(Math.random() * 16777215).toString(16)}`;
+
+    const user = useAppSelector((state) => state.user.currentUser)
+
+    function formatDate(date) {
+        let year = date.getFullYear();
+        let month = date.getMonth() + 1
+        let day = date.getDate()
+        let d = `${year}-${month}-${day}`
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:21]: please use const for variables you dont change

Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:21]: please rename d to understandable name

File: client-side/src/graphs/graphs.jsx [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        return d;
+    }
+
+    let dateStart = new Date(startDate.$d);
+    dateStart = formatDate(dateStart)
+    let dateEnd = new Date(endDate.$d);
+    dateEnd = formatDate(dateEnd)
+    if (websites.loading || users.loading) return <p>Loading...</p>;
+
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:30]: please use loader from storybook

File: client-side/src/graphs/graphs.jsx [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+    if (websites.error || users.error) return <p>Error</p>;
+
+    function getWebsites(dateStart, dateEnd) {
+        let website = websites?.data?.websites?.map((obj) => ({ ...obj, color: getRandomColor() }))
+        const u = users?.data?.users?.find(u => u.email === user.email)
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:35]: replace with understandable name

File: client-side/src/graphs/graphs.jsx [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+
+    let dateStart = new Date(startDate.$d);
+    dateStart = formatDate(dateStart)
+    let dateEnd = new Date(endDate.$d);
+    dateEnd = formatDate(dateEnd)
+    if (websites.loading || users.loading) return <p>Loading...</p>;
+
+    if (websites.error || users.error) return <p>Error</p>;
+
+    function getWebsites(dateStart, dateEnd) {
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:33]: logic should be placed in separated files, for example graphsUtils.js, and be called from the component

File: client-side/src/graphs/graphs.jsx [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+            }
+        });
+
+        return website
+    }
+    let website = getWebsites(dateStart, dateEnd)
+
+
+    const otherProps = {
+        width: 400,
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:70]: please decalre width&height in percents, for responsiveness

File: client-side/src/graphs/graphs.jsx [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+        website.forEach(site => {
+            const websiteName = site.name;
+            if (totalActivityTimeByWebsite[websiteName]) {
+                site.time = totalActivityTimeByWebsite[websiteName];
+            }
+        });
+
+        return website
+    }
+    let website = getWebsites(dateStart, dateEnd)
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:66]: add ;

File: client-side/src/graphs/graphs.jsx [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+        sx: {
+            [`.${ChartsLegend.root}`]: {
+                transform: 'translate(20px, 0)',
+            },
+        },
+    };
+
+    return (
+        <>
+            <h4>this is your data </h4>
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:81]: move texts to constants file,
change text to:
Your browsing history
from {start} to {end}

File: client-side/src/graphs/graphs.jsx [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+            <h4>between the dates {dateStart} and {dateEnd}</h4>
+            <Stack direction="row" width="100%" textAlign="center" spacing={2}>
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:83]: move style declarations to scss file

File: client-side/src/index.js [hunk 27]
```diff
@@ -1,16 +1,33 @@
...
+  cache: new InMemoryCache()
+
+});
 
 const root = ReactDOM.createRoot(document.getElementById('root'));
 root.render(
+
   <React.StrictMode>
-    <App/>
+    <ApolloProvider client={client}>
```
Comment (SariGlick) [inline client-side/src/index.js:26]: does apolloProvider needs to be declare here? Can't it be declared in statistics page only?

File: server-side/.env [hunk 28]
```diff
@@ -1,5 +1,5 @@
-PORT=3000
+PORT=5000
 password=EuACQK5n92JI3vn4
-DB_URL="mongodb://127.0.0.1:27017/TimeOut"
+DB_URL="mongodb+srv://sh3000444:EuACQK5n92JI3vn4@cluster0.2ttcjcg.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"
```
Comment (SariGlick) [inline server-side/.env:3]: remove mongo url change

File: server-side/controllers/profile.controller.js [hunk 29]
```diff
@@ -2,16 +2,15 @@ import mongoose  from 'mongoose';
...
-        const profiles = await Profiles.find().populate('limitedWebsites.websiteId blockedSites').select('-__v');
-        res.json(profiles);
-    } catch (err) {
-        next({message:err.message,status:500})
-    }
-};
 
-  
+   export const  getAllProfiles= async (req, res) => {
+        try {
```
Comment (SariGlick) [inline server-side/controllers/profile.controller.js:7]: Is this change related to your task?

File: server-side/controllers/user.controller.js [hunk 30]
```diff
@@ -1,88 +1,92 @@
-import mongoose  from 'mongoose';
+import mongoose from 'mongoose';
```
Comment (SariGlick) [inline server-side/controllers/user.controller.js:1]: Are the changes in this file related to your task?

File: server-side/models/profile.model.js [hunk 31]
```diff
@@ -7,9 +7,13 @@ const profileSchema = new mongoose.Schema({
         websiteId: { type: Schema.Types.ObjectId, ref: 'Websites' },
         status: { type: String, enum: ['block', 'open'] },
         limitedTimes: [{
-            start: {type:Date,required:true},
-            end: {type:Date,required:true} }]
+            start: { type: Date, default: new Date() },
```
Comment (SariGlick) [inline server-side/models/profile.model.js:10]: Are the changes in this file related to your task?

File: server-side/models/user.model.js [hunk 32]
```diff
@@ -8,7 +8,7 @@ const userSchema = new mongoose.Schema({
   profileImage: { type: String,default:"profile.jpg" },
   visitsWebsites: [{ type: Schema.Types.ObjectId, ref: 'VisitedWebsite' }],
   profiles: [{ type: Schema.Types.ObjectId, ref: 'Profiles' }],
-  preference: { type: Schema.Types.ObjectId, ref: 'Preference' }
+  preferences: { type: Schema.Types.ObjectId, ref: 'Preference' }
```
Comment (SariGlick) [inline server-side/models/user.model.js:11]: Are the changes in this file related to your task?

File: server-side/router/preference.router.js [hunk 33]
```diff
@@ -2,12 +2,10 @@ import  express from 'express';
...
-const preferencesRouter=express.Router();
-
-preferencesRouter.get('/',getAllPreference);
-preferencesRouter.get('/:id', getPreferenceById);
-preferencesRouter.post('/',upload.single('soundVoice'),addPreference);
-preferencesRouter.put('/:id',upload.single('soundVoice'),updatePreference);
-preferencesRouter.delete('/:id',deletePreference);
-
-export default preferencesRouter;
+const router=express.Router();
```
Comment (SariGlick) [inline server-side/router/preference.router.js:5]: Are the changes in this file related to your task?

File: server-side/router/profile.router.js [hunk 34]
```diff
@@ -1,12 +1,9 @@
...
-const profilesRouter=express.Router();
-
-profilesRouter.get('/',getAllProfiles);
-profilesRouter.get('/:id',getProfileById);
-profilesRouter.post('/',createProfile);
-profilesRouter.delete('/:id',deleteProfile);
-profilesRouter.put('/:id',updateProfile);
-
-export default profilesRouter;
+const router=express.Router();
```
Comment (SariGlick) [inline server-side/router/profile.router.js:3]: Are the changes in this file related to your task?

File: server-side/router/user.router.js [hunk 35]
```diff
@@ -2,15 +2,11 @@ import express from 'express';
...
-usersRouter.post('/',upload.single('profileImage'), addUser);
-usersRouter.delete('/:id', deleteUser);
-usersRouter.put('/:id',upload.single('profileImage'), updatedUser);
-
-
-export default usersRouter;
-
-
+const router=express.Router();
+router.get('/users/', getUsers);
```
Comment (SariGlick) [inline server-side/router/user.router.js:6]: Are the changes in this file related to your task?

File: server-side/router/visitedWebsite.router.js [hunk 36]
```diff
@@ -1,12 +1,9 @@
...
-
-visitedWebSitesRouter.get('/',getAllVisitedWebsites)
-visitedWebSitesRouter.get('/:id',getVisitedWebsiteById)
-visitedWebSitesRouter.post('/',createVisitedWebsite)
-visitedWebSitesRouter.put('/:id',updateVisitedWebsite)
-visitedWebSitesRouter.delete('/:id',deleteVisitedWebsite)
-export default visitedWebSitesRouter;
-
+const router=express.Router();
+router.get('/vistedWebsite',getAllVisitedWebsites)
```
Comment (SariGlick) [inline server-side/router/visitedWebsite.router.js:4]: Are the changes in this file related to your task?

File: server-side/router/websites.router.js [hunk 37]
```diff
@@ -1,13 +1,10 @@
...
- const websitesRouter=express.Router();
-
-websitesRouter.get('/',getAllWebsites);
-websitesRouter.get('/:id',getWebsiteById);
-websitesRouter.put('/:id',UpdateWebSite);
-websitesRouter.post('/',addWebSite);
-websitesRouter.delete('/:id',deleteWebsite);
-
-export default websitesRouter;
+ const router=express.Router();
```
Comment (SariGlick) [inline server-side/router/websites.router.js:3]: Are the changes in this file related to your task?

File: statistics-service/graphQL-server.js [hunk 38]
```diff
@@ -0,0 +1,15 @@
...
+const server = new ApolloServer({
+    typeDefs,
+    resolvers,
+})
+
+const { url } = await startStandaloneServer(server, {
+    listen: { port: 4000 },
+});
+
+console.log(`🚀 Server ready at: ${url}`);
```
Comment (SariGlick) [inline statistics-service/graphQL-server.js:15]: remove console.log

File: statistics-service/graphQL-server.js [hunk 39]
```diff
@@ -0,0 +1,15 @@
+import { ApolloServer } from "@apollo/server";
+import { startStandaloneServer } from "@apollo/server/standalone";
```
Comment (SariGlick) [inline statistics-service/graphQL-server.js:2]: join imports form '@apollo/server'

File: statistics-service/reslovers.js [hunk 40]
```diff
@@ -0,0 +1,53 @@
...
+        },
+        users: async () => {
+            try {
+                const response = await axios.get(`${url}:${port}/users/users`);
+                return response.data;
+            } catch (error) {
+                throw new Error(error.message);
+            }
+        },
+        profile: async () => {
```
Comment (SariGlick) [inline statistics-service/reslovers.js:26]: why do you add resolvers for collections you didnt need?

File: statistics-service/schema.js [hunk 41]
```diff
@@ -0,0 +1,60 @@
...
+    end: String
+  }
+
+     type LimitedWebsite {
+    websiteId: Websites
+    status: String
+    limitedTimes: [LimitedTimes]
+  }
+
+  type Profiles {
```
Comment (SariGlick) [inline statistics-service/schema.js:28]: why do you declare types to collections you dont use?

File: client-side/src/stories/DateTime/DateInputs.jsx [hunk 42]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+import PropTypes from 'prop-types';
```
Comment (SariGlick) [inline client-side/src/stories/DateTime/DateInputs.jsx:2]: the datetime isn't part of your task, please take it from existing files in main branch

File: client-side/src/stories/DateTime/DateInputs.scss [hunk 43]
```diff
@@ -0,0 +1,38 @@
+// @import '../stylies.scss';
+
+
```
Comment (SariGlick) [inline client-side/src/stories/DateTime/DateInputs.scss:3]: the datetime isn't part of your task, please take it from existing files in main branch

File: client-side/src/stories/DateTime/DateInputs.stories.js [hunk 44]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+
+import DateInputs from './DateInputs';
```
Comment (SariGlick) [inline client-side/src/stories/DateTime/DateInputs.stories.js:3]: the datetime isn't part of your task, please take it from existing files in main branch

**Review by SariGlick — CHANGES_REQUESTED** [review]

File: client-side/src/graphs/graphs.jsx [hunk 45]
```diff
@@ -0,0 +1,59 @@
+import * as React from 'react';
```
Comment (SariGlick) [inline client-side/src/graphs/graphs.jsx:1]: please move all graphs files to be under components/statistics

**Review by SariGlick — CHANGES_REQUESTED** [review]

**Review by MalkySinowitz — APPROVED** [review]
looks great

File: server-side/controllers/profile.controller.js (hunk 29 above)
Comment (e171271) [reply server-side/controllers/profile.controller.js]: you wrote me in slack that it's not related to you, so why do you have this changes?

**Review by SariGlick — APPROVED** [review]

**Review by SariGlick — APPROVED** [review]
